### PR TITLE
Add ListenerFactory + Test coverage 100% + MSI 100% + Psalm 1 + Minor improvements

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

**Comment by samdark** [comment]
I don't like that now we are creating every listener two times in debug mode.

File: src/ListenerFactory.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+use ReflectionMethod;
+
+use function is_array;
+use function is_callable;
+use function is_string;
+
+/**
+ * Create real callable listener from configuration.
+ */
+final class ListenerFactory
```
Comment (viktorprogger) [inline src/ListenerFactory.php:18]: Shouldn't it be `@internal`?

File: src/ListenerConfigurationChecker.php [hunk 2]
```diff
@@ -69,41 +67,17 @@ public function check(array $configuration): void
...
-
-            if ($this->container->has($definition[0])) {
-                $object = $this->container->get($definition[0]);
-
-                return method_exists($object, $definition[1]);
-            }
-
+        try {
+            $this->listenerFactory->create($definition);
+        } catch (InvalidListenerConfigurationException $e) {
```
Comment (viktorprogger) [inline src/ListenerConfigurationChecker.php:77]: It shouldn't be caught as it is caught again and converted to `InvalidListenerConfigurationException` at a higher level (the `check` method). We can just replace `if (!$this->isCallable...` to `$this->listenerFactory->create($definition);`

**Review by viktorprogger — CHANGES_REQUESTED** [review]

File: src/ListenerFactory.php (hunk 1 above)
Comment (vjik) [reply src/ListenerFactory.php]: yes, done.

**Comment by vjik** [comment]
> I don't like that now we are creating every listener two times in debug mode.

- in checker we create `callable`
- in collection we create and invoke `callable` only on fire event

Renamed `ListenerFactory` to `ListenerCallableFactory`.

File: src/ListenerConfigurationChecker.php (hunk 2 above)
Comment (vjik) [reply src/ListenerConfigurationChecker.php:79]: Factory throws two exceptions:
- `InvalidListenerConfigurationException ` on failed create listener, except for internal error in container;
- `ContainerExceptionInterface` on error while retrieving the entry from container.

The `check ` method catches only `ContainerExceptionInterface`, and here we catch `InvalidListenerConfigurationException ` for create more detailed error message.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**